sled-agent/src/sim/storage.rs

```diff
@@ -1009,7 +1099,149 @@ impl Storage {
         })
     }
 
-    pub async fn omicron_physical_disks_list(
+    pub fn nested_dataset_list(
+        &self,
+        name: NestedDatasetLocation,
+        options: NestedDatasetListOptions,
+    ) -> Result<Vec<NestedDatasetConfig>, HttpError> {
+        let Some(mut nested_dataset) = self.nested_datasets.get(&name.root)
+        else {
+            return Err(HttpError::for_not_found(
+                None,
+                "Dataset not found".to_string(),
+            ));
+        };
+
+        for path_component in name.path.split('/') {
+            if path_component.is_empty() {
+                continue;
+            }
+            match nested_dataset.children.get(path_component) {
+                Some(dataset) => nested_dataset = dataset,
+                None => {
+                    return Err(HttpError::for_not_found(
+                        None,
+                        "Dataset not found".to_string(),
+                    ))
+                }
+            };
+        }
+
+        let mut children: Vec<_> = nested_dataset
+            .children
+            .values()
+            .map(|storage| storage.config.clone())
+            .collect();
+
+        match options {
+            NestedDatasetListOptions::ChildrenOnly => return Ok(children),
+            NestedDatasetListOptions::SelfAndChildren => {
+                children.insert(0, nested_dataset.config.clone());
+                return Ok(children);
+            }
+        }
+    }
+
+    pub fn nested_dataset_ensure(
+        &mut self,
+        config: NestedDatasetConfig,
+    ) -> Result<(), HttpError> {
+        let name = &config.name;
+        let nested_path = name.path.to_string();
+        let zpool_root = self.root().to_path_buf();
+        let Some(mut nested_dataset) = self.nested_datasets.get_mut(&name.root)
+        else {
+            return Err(HttpError::for_not_found(
+                None,
+                "Dataset not found".to_string(),
+            ));
+        };
+
+        for path_component in nested_path.split('/') {
+            if path_component.is_empty() {
+                continue;
+            }
+
+            // Final component of path -- insert it here if it doesn't exist
+            // already.
+            if !path_component.contains('/') {
+                let entry =
+                    nested_dataset.children.entry(path_component.to_string());
+                entry
+                    .and_modify(|storage| {
+                        storage.config = config.clone();
+                    })
+                    .or_insert_with(|| {
+                        NestedDatasetStorage::new(
+                            &zpool_root,
+                            config.name.root,
+                            nested_path,
+                            config.inner,
+                        )
+                    });
+                return Ok(());
+            }
```

Comment:
This condition will always evaluate to true since we're splitting `path_component` on `/`, so we'll never walk the full path.


https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=530576a88d3a03606274a9d488462528